Extend `YesodAuthEmail` to support extensible password hashing

bigs opened the pull request:
## Commit Log

This change introduces `hashAndSaltPassword` and `verifyPassword` to the
`YesodAuthEmail` type class, allowing users to implement their own hashing
schemes (i.e. to provide compatibility with an existing database). It also
updates the default handlers to use these new functions when appropriate. The
functions have default implementation such that behavior for legacy applications
should not change.

## Notes

### Prior Art

The only other discussion about offering extensible password hashing was found in [this email list thread](https://groups.google.com/forum/#!topic/yesodweb/dqZDA18Qjy8). It didn't seem to come to a conclusion, though there have been others asking about this functionality.

### Regressions

There are currently no tests in the `yesod-auth` project so there was nothing to measure a regression against. I am, however, successful in using in a project I'm currently working on that requires integration with a legacy database.

### Steps to Reproduce (stack)

Clone this revision locally and add the `yesod/yesod-auth` folder to your `packages` list in your `stack.yaml`. Build your project using `YesodAuthEmail` and the behavior should be no different from v1.4.19.

Should you choose to implement an alternative password hashing, provide definitions for `hashAndSaltPassword` and `verifyPassword`.

### Miscellanea

Happy to rename or reformat any functions, I'm not married to anything here.

psibi commented on yesod-auth/Yesod/Auth/Email.hs:
```diff
@@ -203,6 +203,18 @@ class ( YesodAuth site
+    -- | Hash and salt a password
+    --
+    -- Default: 'saltPass'.
+    hashAndSaltPassword :: Text -> HandlerT site IO SaltedPass
```
Add @since haddock style documentation for all the newly introduced functions.

psibi commented on yesod-auth/Yesod/Auth/Email.hs:
```diff
@@ -756,11 +770,14 @@ postPasswordR = do
-                        | otherwise ->
-                            lift $ loginErrorMessage (tm setpassR) "Invalid current password, please try again"
+                    Just realpass -> do
+                        passValid <- lift $ verifyPassword (fromJust current) realpass
```
Avoid using partial functions.

bigs commented:
@psibi thanks! the partial function was lifted from and i knew i should have handled it there! not entirely sure what you mean re: haddock, though. could you explain?

bigs commented:
@psibi i've removed the partial function and added what seems like an appropriate redirection for the case (which afaict shouldn't occur). if the user didn't supply a password, i redirect them to the login page with an appropriate message.

bigs commented:
@psibi ohhhhhhhhhhhhh github was just auto-correcting `@since` to a username. one thing i'm not sure about is whether or not i should bump the version number. should i go to `1.4.20`? with that answer in, i'll update the haddocks!

psibi commented on yesod-auth/Yesod/Auth/Email.hs:
```diff
@@ -753,14 +771,20 @@ postPasswordR = do
+                        passValid <- lift $ verifyPassword current' realpass
+                        case () of
+                            _
+                                | isNothing current -> loginErrorMessageI LoginR Msg.BadSetPass
```
Can this check be avoided as we are doing this on top .also ?

bigs replied:
ahhh great point. in fact, this is the proper error message to deliver. refactoring & updating version

psibi commented:
@bigs Also it should be `1.4.20`

bigs commented:
@psibi refactors and version bump added as requested!

bigs commented:
sounds good! backwards compatible, yes. will squash and update log!

bigs commented:
@psibi changelog updated and branch squashed :) should be set!

psibi commented:
Thanks for the PR!

bigs commented:
@psibi happy to contribute! hoping it's the first of many.
